src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java

```diff
@@ -0,0 +1,105 @@
+package io.vertx.rabbitmq;
+
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.BasicProperties;
+import io.vertx.codegen.annotations.GenIgnore;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.rabbitmq.impl.RabbitMQPublisherImpl;
+
+/**
+ * A reliable publisher that
+ * <ul>
+ * <li>Queues up messages internally until it can successfully call basicPublish.
+ * <li>Notifies the caller using a robust ID (not delivery tag) when the message is confirmed by rabbit.
+ * </ul>
+ * 
+ * This is a layer above the RabbitMQClient that provides a lot of standard implementation when guaranteed at least once delivery is required.
+ * If confirmations are not required do not use this publisher as it does have overhead.
+ * 
+ * @author jtalbut
+ */
+@VertxGen
+public interface RabbitMQPublisher { 
+  
+  /**
+   * Create and return a publisher using the specified client.
+   *
+   * @param vertx  the vertx instance.
+   * @param client the RabbitMQClient.
+   * @param options options for the publisher.
+   * @param connectionEstablishedCallback callback to be called each time a connection is established (it is usually a good idea to declare the exchange in this callback).
+   * @return the publisher
+   */
+  static RabbitMQPublisher create(Vertx vertx
+          , RabbitMQClient client
+          , RabbitMQPublisherOptions options
+  ) {
+    return new RabbitMQPublisherImpl(vertx, client, options);
+  }
+  
+  /**
+   * Start the rabbitMQ publisher.
+   * The RabbitMQClient should have been started before this.
+   *
+   */
+  void start(Handler<AsyncResult<Void>> resultHandler);
+  
+  /**
+   * Like {@link #start(Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<Void> start();
+  
+  /**
+   * Stop the rabbitMQ publisher.
+   * Calling this is optional, but it gives the opportunity to drain the send queue without losing messages.
+   * Future calls to publish will be ignored.
+   *
+   */
+  void stop(Handler<AsyncResult<Void>> resultHandler);
+  
+  /**
+   * Like {@link #stop(Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<Void> stop();
+  
+  /**
+   * Undo the effects of calling {@link #stop(Handler)} so that publish may be called again. 
+   */
+  void restart();
```

Comment:
what does restart do if stop() is not called ?

Reply:
It's harmless to call restart when stop hasn't been called (it'll pause and resume the stream, that's all).
However, if restart is called when stop has been called but has not yet completed then the stop will never complete.
I've added a comment to say that, do you think it needs more?